cpp/dolfinx/common/subsystem.cpp

```diff
@@ -64,18 +63,26 @@ void subsystem::init_logging(int argc, char* argv[])
   loguru::init(argc, argv_copy.data(), options);
 }
 //-----------------------------------------------------------------------------
-void subsystem::init_petsc()
+void subsystem::init_petsc(MPI_Comm comm)
 {
   int argc = 0;
   char** argv = nullptr;
-  init_petsc(argc, argv);
+  init_petsc(argc, argv, comm);
 }
 //-----------------------------------------------------------------------------
-void subsystem::init_petsc(int argc, char* argv[])
+void subsystem::init_petsc(int argc, char* argv[], MPI_Comm comm)
 {
   if (argc > 1)
     LOG(INFO) << "Initializing PETSc with given command-line arguments.";
 
+  // Ensure MPI has been initialised, otberwise setting PETSC_COMM_WORLD can 
+  // call MPI not initialised errors.
+  init_mpi(argc, argv);
```

Comment:
In the interests of being explicit, I would suggest checking that MPI has been initialised and throwing and error if it hasn't, and suggest that MPI needs to be initialised by the user first. Risk at the moment is confusion around who is responsible for finalising MPI.

Reply:
Good point. The reason I didn't do that is that `PetscInitialize` also will implicitly initialise MPI if it's not initialised. If you're happy moving to the model that users always have to initialise MPI first before calling PETSc (which would be a change to the current behaviour, but would be much more explicit) then that's fine.

Reply:
Checking `PetscFinalize` I see it will only call `MPI_Finalize` if it also called `MPI_Init` but will not call `MPI_Finalize` if the user called `MPI_Init` prior to calling `PetscInitialize` (i.e. `init_petsc`)